### ISSUE TITLE
Weird compiler crash triggered when using `uint` as a parameter

### ISSUE THREAD

**Issue by fxzjshm** [google/clspv#760]
Hi! Recently I encountered a weird crash of clspv, the simplified code is
```OpenCL
__kernel void shift(__global int *in, __global int *out, uint t0) {
    uint t = get_global_id(0);
    uint x;
    if(t > t0){
        x = t - t0;
    } else {
        x = 0;
    }
    out[t] = in[x];
}
```
clspv crashes when compiling this code using `clspv temp.cl -o temp.spv -O1` (or `-O2`, `-O3`), the output is
```
; Function Attrs: nofree nosync nounwind readnone speculatable willreturn
declare i32 @llvm.usub.sat.i32(i32, i32) #1
[1]    14310 segmentation fault (core dumped)  clspv temp.cl -o temp.spv -O1
```
exit code is 139

Here's what I've tried:
* When using`-O0` the error disappears
* If I change `uint t0` to `int t0` the error disappears
* the type of `t`, `x`, `in`, `out` seems not affecting the result (but not sure)
* If comment out `out[t] = in[x]` the error disappears
* moving `if` statment into a separate function doesn't work
  * ```OpenCL
    inline uint get_index(uint t, uint t0) {
        uint x;
        if(t > t0){
            x = t - t0;
        } else {
            x = 0;
        }
        return x;
    }

    __kernel void shift(__global uint *in, __global uint *out, uint t0) {
        uint t = get_global_id(0);
        uint x = get_index(t, t0);
        out[t] = in[x];
    }
    ```

I don't know whether it's a limitation of clspv or there's some bug in it. I'm not familiar with llvm either, thus don't know how to debug this, could you help me?

**Comment by alan-baker** [comment]
This looks like clspv is not expecting the saturated subtraction. If you want to work around this, you should be able to use the OpenCL C `sub_sat` intrinsic directly, e.g.:
```
inline uint get_index(uint t, uint t0) {
    uint x = sub_sat(t, t0);
    return x;
}
```

The fix is to commonize how `sub_sat` is handled to also handle `llvm.usub.sat`.

**Comment by fxzjshm** [comment]
Sorry for the very late response, I have been busy studying all this semester and forgot this issue.
Thank you for the workaround, confirm it works.
But I'm totally unfamiliar with LLVM, so after some investigation I didn’t find where to add the handling of `llvm.usub.sat`.

**Comment by alan-baker** [comment]
The code in ReplaceOpenCLBuiltin should generate the llvm intrinsic and the lowering that used to happen there (or in SPIRVProducerPass) should instead be moved into the ReplaceLLVMIntrinsics so that it happens more generally. 

An example to follow would be to look at the work that happened on #682. Happy to help more if you need it. I appreciate if you can post a patch to fix this.

**Comment by gnl21** [comment]
Right, that makes sense. Thanks. I'll have a look at producing a fix.
